### PR TITLE
fix: [PL-41096]: Add taskId for Sm's validation task

### PR DESCRIPTION
**Pull request by VikasMaddukuriHarness**
## Describe your changes
Add taskId to the secret manager validation task. 
For all the other connectors which are not of type SM, recently (a couple month ago) pr was merged which gives back the taskId when testConnection is done for a connector.
That pr missed the SM type of connectors.
earlier pr - https://github.com/harness/harness-core/pull/46630 
This pr is to fix the SM type connectors.

Testing:
1. Tested in local (both with new delegate(develop) and old delegate 803xx delegate) and both worked fine. 
Attaching screenShots of some Sm's testConnections
HashiCorp vault -
Screenshot 2023-09-06 at 3 58 05 PM

AzureKeyVault - 
Screenshot 2023-09-06 at 3 58 42 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51443)

### CODE REVIEW

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGSecretManagerServiceImpl.java [hunk 1]
```diff
@@ -125,7 +126,7 @@ public boolean validateNGSecretManager(
                 validationResult = vaultEncryptorsRegistry.getVaultEncryptor(VAULT).validateSecretManagerConfiguration(
                     accountIdentifier, vaultConfig);
               } else {
-                validationResult = true;
+                validationResult = Pair.of("", true);
```
Comment (xingchi-jin) [inline 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGSecretManagerServiceImpl.java:129]: if the task doesn't happen, can we not return empty, create a constant string, like "NO_TASK_ID" .

File: 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGSecretManagerServiceImpl.java [hunk 2]
```diff
@@ -197,19 +198,24 @@ public ConnectorValidationResult validate(
       @NotNull String accountIdentifier, String orgIdentifier, String projectIdentifier, @NotNull String identifier) {
     ConnectivityStatus connectivityStatus = ConnectivityStatus.FAILURE;
     SecretManagerConfigDTO secretManagerConfigDTO = null;
+    String taskId = "";
     try {
       secretManagerConfigDTO = ngConnectorSecretManagerService.getUsingIdentifier(
           accountIdentifier, orgIdentifier, projectIdentifier, identifier, false);
-      if (validateNGSecretManager(accountIdentifier, secretManagerConfigDTO)) {
+      var responseEntry = validateNGSecretManager(accountIdentifier, secretManagerConfigDTO);
+      if (responseEntry.getValue()) {
```
Comment (xingchi-jin) [inline 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/NGSecretManagerServiceImpl.java:206]: does this mean task id is only set when sm connector validation succeeds ? how about when the task fails ?

File: 400-rest/src/main/java/io/harness/encryptors/managerproxy/ManagerEncryptorHelper.java [hunk 3]
```diff
@@ -155,7 +157,7 @@ public boolean validateConfiguration(String accountId, ValidateSecretManagerConf
       }
       ValidateSecretManagerConfigurationTaskResponse responseData =
           (ValidateSecretManagerConfigurationTaskResponse) delegateResponseData;
-      return responseData.isConfigurationValid();
+      return Pair.of("", responseData.isConfigurationValid());
```
Comment (xingchi-jin) [inline 400-rest/src/main/java/io/harness/encryptors/managerproxy/ManagerEncryptorHelper.java:160]: looks like it always return empty task id, am i missing something ?

File: 970-api-services-beans/src/main/java/io/harness/encryptors/KmsEncryptor.java [hunk 4]
```diff
@@ -24,7 +25,7 @@ EncryptedRecord encryptSecret(
   char[] fetchSecretValue(
       @NotEmpty String accountId, @NotNull EncryptedRecord encryptedRecord, @NotNull EncryptionConfig encryptionConfig);
 
-  default boolean validateKmsConfiguration(String accountId, EncryptionConfig encryptionConfig) {
+  default Pair<String, Boolean> validateKmsConfiguration(String accountId, EncryptionConfig encryptionConfig) {
```
Comment (xingchi-jin) [inline 970-api-services-beans/src/main/java/io/harness/encryptors/KmsEncryptor.java:28]: can you add a new function called validateKmsConfigurationReturnTaskId and not to change the interface of this function.

Let's avoid changing codes that touches delegate, like codes in 931 . Those encryptors are not creating delegate tasks, the codes of them should not change at all

File: 400-rest/src/main/java/io/harness/encryptors/managerproxy/ManagerEncryptorHelper.java (hunk 3 above)
Comment (xingchi-jin) [reply 400-rest/src/main/java/io/harness/encryptors/managerproxy/ManagerEncryptorHelper.java:160]: ignore the above comment, I initially couldn't understand the code completely

But pl. check the next comment. If managerEncryptor doesn't create delegate task, let's not change the code

File: 970-api-services-beans/src/main/java/io/harness/encryptors/KmsEncryptor.java (hunk 4 above)
Comment (VikasMaddukuriHarness) [reply 970-api-services-beans/src/main/java/io/harness/encryptors/KmsEncryptor.java:28]: Basically the interface that we are using a interface which being used for implementing all the encryptors and the manager and ngmanager encryptor. So if we were to modify the method signature in manager and ng-manager encryptor, we need to modify the interface signature and this inturns requires all the encryptors method signature.

If we want to decouple them, that would have more code changes then.
But I did test the back ward compatibility by using an old delegate and the code changes looks fine. 
Let me know if you want to discuss this. Thanks.

Comment (xingchi-jin) [reply 970-api-services-beans/src/main/java/io/harness/encryptors/KmsEncryptor.java:28]: is it possible to do like this
`
interface KmsEncryptor {
   EncryptedRecord encryptSecret(
      @NotEmpty String accountId, @NotEmpty String value, @NotNull EncryptionConfig encryptionConfig);

  char[] fetchSecretValue(
      @NotEmpty String accountId, @NotNull EncryptedRecord encryptedRecord, @NotNull EncryptionConfig encryptionConfig);

  default boolean validateKmsConfiguration(String accountId, EncryptionConfig encryptionConfig) {
    throw new UnsupportedOperationException(
        "Validating SecretManager Configuration on Delegate in not available yet for:" + encryptionConfig);
  }
}

interface NgCgManagerKmsEncryptor extends  KmsEncryptor {
  pair<String, Boolean> validateKmsConfigurationReturnTaskId(String accountId, EncryptionConfig encryptionConfig);
}

then, let manager and ng-manager encryptor implement the NgCgManagerKmsEncryptor and let the actual encryptor implement the KmsEncryptor 
`

**Pull request closed without merging**